src/message-io.js

```diff
@@ -1,74 +1,22 @@
 const tls = require('tls');
 const crypto = require('crypto');
 const EventEmitter = require('events').EventEmitter;
-const Transform = require('readable-stream').Transform;
 
-const Packet = require('./packet').Packet;
+const IncomingMessageStream = require('./message/incoming-message-stream');
+const OutgoingMessage = require('./message/outgoing-message');
 const TYPE = require('./packet').TYPE;
 const packetHeaderLength = require('./packet').HEADER_LENGTH;
 
-class ReadablePacketStream extends Transform {
-  constructor() {
-    super({ objectMode: true });
-
-    this.buffer = new Buffer(0);
-    this.position = 0;
-  }
-
-  _transform(chunk, encoding, callback) {
-    if (this.position === this.buffer.length) {
-      // If we have fully consumed the previous buffer,
-      // we can just replace it with the new chunk
-      this.buffer = chunk;
-    } else {
-      // If we haven't fully consumed the previous buffer,
-      // we simply concatenate the leftovers and the new chunk.
-      this.buffer = Buffer.concat([
-        this.buffer.slice(this.position), chunk
-      ], (this.buffer.length - this.position) + chunk.length);
-    }
-
-    this.position = 0;
-
-    // The packet header is always 8 bytes of length.
-    while (this.buffer.length >= this.position + packetHeaderLength) {
-      // Get the full packet length
-      const length = this.buffer.readUInt16BE(this.position + 2);
-
-      if (this.buffer.length >= this.position + length) {
-        const data = this.buffer.slice(this.position, this.position + length);
-        this.position += length;
-        this.push(new Packet(data));
-      } else {
-        // Not enough data to provide the next packet. Stop here and wait for
-        // the next call to `_transform`.
-        break;
-      }
-    }
-
-    callback();
-  }
-}
-
 module.exports = class MessageIO extends EventEmitter {
   constructor(socket, _packetSize, debug) {
     super();
 
     this.socket = socket;
     this._packetSize = _packetSize;
     this.debug = debug;
-    this.sendPacket = this.sendPacket.bind(this);
-
-    this.packetStream = new ReadablePacketStream();
-    this.packetStream.on('data', (packet) => {
-      this.logPacket('Received', packet);
-      this.emit('data', packet.data());
-      if (packet.isLast()) {
-        this.emit('message');
-      }
-    });
 
-    this.socket.pipe(this.packetStream);
+    this.incomingMessageStream = new IncomingMessageStream(this.debug);
```

Comment:
> This was the "simplest" way to implement this without shuffling too many things around

Doing a "new IncomingMessageStream(this.debug)" in Connection class and passing it to new MessageIO() shouldn't shuffle too many things :-)

Agree on the longer term, MessageIO does seem like a Duplex stream to me.